app/controllers/concerns/idv/doc_auth_vendor_concern.rb

```
@@ -7,19 +7,29 @@ module DocAuthVendorConcern
     # @returns[String] String identifying the vendor to use for doc auth.
     def doc_auth_vendor
       if resolved_authn_context_result.facial_match? || socure_user_set.maxed_users?
-        if doc_auth_vendor_enabled?(Idp::Constants::Vendors::LEXIS_NEXIS)
-          bucket = :lexis_nexis
-        elsif doc_auth_vendor_enabled?(Idp::Constants::Vendors::MOCK)
-          bucket = :mock
-        else
-          return nil
-        end
+        bucket = choose_non_socure_bucket
       else
         bucket = ab_test_bucket(:DOC_AUTH_VENDOR)
       end
+
+      if bucket == :socure
+        if !add_user_to_socure_set
```

One other thing that occurred to me as a potential edge case. If a user:

1. Ends up in the Socure bucket
2. Is successfully added to the Set, and is the last valid user
3. If they restart IdV, they will not be redirected Socure again (and same for any other users that previously would have been)?

Is that expected?

---

@mitchellhenke, Paired with jmax, alex, and amir, we discuss. This is the expected behavior, once the user has been bucketed, they should stay in the bucket based on uuid even after the try again option is selected. The bucket is selected at the beginning of an idv session. Jmax suggested switching to a lua script due to its atomic nature, going to look into implementing that this morning, but in terms of behavior I think the current solution is sufficient. LMK if you would like to discuss further!

---

I haven't looked closely enough yet, but I suspect there's an approach that wouldn't require a Lua script.

---

was able to move forward with the lua approach, LMK your thoughts. We can revert to the old approach if needed

---

@mitchellhenke - I wasn't able to identify one after looking pretty hard.

The essential problem is that we want the check for max users and add of the user if there is room to be an atomic operation, with logic inside it.
- Redis transactions are atomic, but they don't let us make decisions inside them. They aren't full-blown DB transactions, so the option of rolling back if we exceed the limit isn't there either.
- Lua scripts _are_ atomic; that's exactly why we use them in the rate limiter code.

I'm totally down for a non-script solution; I just don't think there is one.

---

Thanks, I missed the requirement on strict atomicity and whether this was a harder cap or if some slippage was permitted.

If a user is in the Socure bucket, is it expected that they're able to restart doc auth with Socure if the cap has been reached in the intervening time?

---

No. If a user who has previously been bucketed to socure starts doc auth over to the point we need to re-choose a bucket and the max_users have been reached, then they should be places in the lexis_nexis bucket

---

Point. I think that means that @AShukla-GSA's suggestion (while we were pairing) that we actually return whether the user is actually in the set is the right answer. So the *entire* logic in the LUA script becomes:

```
if (we haven't hit the rate cap)
  add the user
end
return whether the user is in the set.
```

Which covers the case where the user is already in the set and we've hit the rate cap, and is slightly simpler to boot.

---

Apologies, I'm not sure I'm following. If someone is assigned the Socure bucket, they are expected to go through Socure regardless of whether the limit is hit?

---

The expected behavior is to make all users default to mock/LexisNexis doc auth vendor once the limit is reached regardless if they have been placed in a socure bucket or not. https://cm-jira.usa.gov/browse/LG-15275 has this in its implementation details. This is how it is currently implemented. I can change the logic to allow users who are already in the set to be bucketed to socure, but this is not the expected behavior.

---

Got it, thanks